### PR TITLE
sagemath-mpmath

### PR DESCRIPTION
**Pull request by mkoeppe**
### :books: Description

Instead of relying on [mpmath's mechanism to select a backend](https://github.com/mpmath/mpmath/blob/master/docs/setup.rst#mpmath-under-sage), which is problematic when other libraries use mpmath too (#25445), we create an isolated copy of mpmath in `sage.libs.mpmath._vendor.mpmath`. The copy is created at build/sdist time using the pypa [vendoring](https://pypi.org/project/vendoring/) tool; we do not put a copy of the sources into our git source tree.

All uses of mpmath in the Sage library are updated. The Sage library no longer imports from `mpmath` and no longer sets environment variables that force `mpmath` (when imported by another library) to use the Sage backend.

The library is intended to be shipped by the pip-installable distribution **sagemath-mpmath** (added here on the PR), which only has **sagemath-categories** as a Python dependency.
Downstream projects that wish to use this version of mpmath should declare a dependency (install-requires) on sagemath-mpmath and use `import sage.libs.mpmath.all` instead of `import mpmath`.

- Fixes #25445
- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by dimpase** [comment]
what kinds of simplifications could be done on `mpmath` side here - assuming this is merged?

**Comment by mkoeppe** [comment]
> what kinds of simplifications could be done on `mpmath` side here - assuming this is merged?

I would suggest that they remove the condition `'SAGE_ROOT' in os.environ` from https://github.com/mpmath/mpmath/blob/master/mpmath/libmp/backend.py#L41 (in fact, this could be done already now because Sage has not depended on this mechanism since #32799.

Other than that, I don't think anything gets simpler for mpmath itself; but for users of mpmath and Sage, the situation becomes clearer and more robust.

**Comment by dimpase** [comment]
before doing anything here, we should check whether just getting rid of Sage backend in mpmath (which is a very old hack) is the right thing to do. I.e. whether there is a serious speed regression, something I doubt.

**Comment by mkoeppe** [comment]
> we should check whether just getting rid of Sage backend in mpmath (which is a very old hack) is the right thing to do.

+1. Do you know how to benchmark it, or should we get help from upstream?

**Comment by dimpase** [comment]
it's a mixed bag - some tests are slower, some are faster. E.g. here are 5 slowest tests from
[mpmath](https://github.com/mpmath/mpmath/tree/88f26061a5f6c6350f45e69dacf871547c97696b)/[mpmath](https://github.com/mpmath/mpmath/tree/88f26061a5f6c6350f45e69dacf871547c97696b/mpmath)/[tests](https://github.com/mpmath/mpmath/tree/88f26061a5f6c6350f45e69dacf871547c97696b/mpmath/tests/test_torture.py)
```
(sage-sh) dima@hilbert:sage-dev$ MPMATH_NOSAGE=yes pytest test_torture.py --durations=5
==================================================================== test session starts =====================================================================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.0.0
rootdir: /mnt/opt/Sage/sage-dev
collected 90 items                                                                                                                                           

test_torture.py ..........................................................................................                                             [100%]

==================================================================== slowest 5 durations =====================================================================
60.31s call     test_torture.py::test_asymp[<lambda>-150-False23]
12.87s call     test_torture.py::test_asymp[<lambda>-150-False9]
12.16s call     test_torture.py::test_asymp[<lambda>-90-False]
10.97s call     test_torture.py::test_asymp[<lambda>-150-False37]
10.76s call     test_torture.py::test_asymp[f_wrapped-90-False]
=============================================================== 90 passed in 286.13s (0:04:46) ===============================================================
(sage-sh) dima@hilbert:sage-dev$ pytest test_torture.py --durations=5
==================================================================== test session starts =====================================================================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.0.0
rootdir: /mnt/opt/Sage/sage-dev
collected 90 items                                                                                                                                           

test_torture.py ..........................................................................................                                             [100%]

==================================================================== slowest 5 durations =====================================================================
59.20s call     test_torture.py::test_asymp[<lambda>-150-False23]
12.18s call     test_torture.py::test_asymp[<lambda>-150-False9]
10.12s call     test_torture.py::test_asymp[<lambda>-90-False]
9.85s call     test_torture.py::test_asymp[<lambda>-150-False37]
9.53s call     test_torture.py::test_asymp[barnesg-90-False]
=============================================================== 90 passed in 239.50s (0:03:59) ===============================================================
```

So overall with Sage backend it's about 20% faster. OTOH one test there is must faster without Sage backend:
```
(sage-sh) dima@hilbert:sage-dev$ MPMATH_NOSAGE=yes pytest test_torture.py -k test_bernoulli_huge
==================================================================== test session starts =====================================================================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.0.0
rootdir: /mnt/opt/Sage/sage-dev
collected 90 items / 89 deselected / 1 selected                                                                                                              

test_torture.py .                                                                                                                                      [100%]

============================================================== 1 passed, 89 deselected in 0.32s ==============================================================
(sage-sh) dima@hilbert:sage-dev$ pytest test_torture.py -k test_bernoulli_huge
==================================================================== test session starts =====================================================================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.0.0
rootdir: /mnt/opt/Sage/sage-dev
collected 90 items / 89 deselected / 1 selected                                                                                                              

test_torture.py .                                                                                                                                      [100%]

============================================================== 1 passed, 89 deselected in 4.08s ==============================================================
```

File: build/pkgs/vendoring/dependencies [hunk 1]
```diff
@@ -0,0 +1,4 @@
+$(PYTHON) rich click toml | $(PYTHON_TOOLCHAIN)
```
Comment (jhpalmieri) [inline build/pkgs/vendoring/dependencies]: There is no `toml` package in Sage. Should this be `tomli` instead?

Comment (mkoeppe) [reply build/pkgs/vendoring/dependencies]: the toml package was removed from Sage in the meantime; we'll have to re-add it for sagemath-mpmath

File: build/pkgs/rich/dependencies [hunk 2]
```diff
@@ -0,0 +1,4 @@
+$(PYTHON) markdown_it_py | $(PYTHON_TOOLCHAIN)
```
Comment (jhpalmieri) [inline build/pkgs/rich/dependencies:1]: This seems to require `poetry-core`?

Comment (mkoeppe) [reply build/pkgs/rich/dependencies:1]: Not any more, I've now made it a wheel package, so there's no more build dependency.

**Comment by mkoeppe** [comment]
partially squashed

**Comment by skirpichev** [comment]
@dimpase, I assume you did tests without gmpy2.  Is there advantages of using sage backend c.f. gmpy?

> we should check whether just getting rid of Sage backend in mpmath

Which will be a simplest solution for mpmath/mpmath#705...

**Comment by mkoeppe** [comment]
Closing this PR as this change appears to be slated for inclusion in mpmath 1.4:
- https://github.com/mpmath/mpmath/pull/732

**Pull request closed without merging**